examples/keyvalue-inmemory/main.go

```diff
@@ -69,12 +71,24 @@ func run() error {
 
 func (p *Provider) handleNewSourceLink(link provider.InterfaceLinkDefinition) error {
 	log.Println("Handling new source link", link)
+	err := p.establishSourceLink(link)
+	if err != nil {
+		log.Println("Failed to establish source link", link, err)
+		p.failedSourceLinks[link.Target] = link
+		return err
+	}
 	p.sourceLinks[link.Target] = link
 	return nil
 }
 
 func (p *Provider) handleNewTargetLink(link provider.InterfaceLinkDefinition) error {
 	log.Println("Handling new target link", link)
+	err := p.establishTargetLink(link)
+	if err != nil {
+		log.Println("Failed to establish target link", link, err)
+		p.failedTargetLinks[link.SourceID] = link
```

Comment:
Do we also need to make sure we delete the link out of the "failed" map during the delete link handlers?

Reply:
I believe I did add the functionality in the below `handleDel...` funcs. 

```
func (p *Provider) handleDelSourceLink(link provider.InterfaceLinkDefinition) error {
	log.Println("Handling del source link", link)
	delete(p.sourceLinks, link.Target)
	return nil
}
```

Unless I'm misunderstanding the question??

Reply:
Ah I see, yeah I see that we delete the link if it was a successful link, but I was wondering if you also needed a line like this to also make sure it's removed from the "failed" map
```go
delete(p.failedSourceLinks, link.Target)
```

Reply:
Yes, you are correct, sorry I misunderstood prior. I will update.